src/Workspaces/Core/Portable/Serialization/IsolatedAnalyzerReferenceSet.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#if NET
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Runtime.Loader;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Host;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Serialization;
+
+/// <summary>
+/// A set of <see cref="IsolatedAnalyzerReference"/>s and their associated <see cref="AssemblyLoadContext"/> (ALC).  As
+/// long as something is keeping this set alive, the ALC will be kept alive.  Once this set is dropped, the ALC will be
+/// explicitly <see cref="AssemblyLoadContext.Unload"/>'ed in its finalizer.
+/// </summary>
+internal sealed class IsolatedAssemblyReferenceSet
+{
+    /// <summary>
+    /// Final set of <see cref="AnalyzerReference"/> instances that will be passed through the workspace down to the compiler.
+    /// </summary>
+    private readonly ImmutableArray<IsolatedAnalyzerReference> _analyzerReferences;
+
+    /// <summary>
+    /// Dedicated ALC that all analyzer references will load their <see cref="System.Reflection.Assembly"/>s within.
+    /// </summary>
+    private readonly AssemblyLoadContext _assemblyLoadContext;
+
+    public ImmutableArray<AnalyzerReference> AnalyzerReferences => _analyzerReferences.CastArray<AnalyzerReference>();
+
+    public IsolatedAssemblyReferenceSet(
+        ImmutableArray<AnalyzerReference> serializedReferences,
+        IAnalyzerAssemblyLoaderProvider provider)
+    {
+        // We should really only be handed SerializedAnalyzerReference here (as that's the mainline case for a host
+        // communicating references to the OOP side).  However, we may also get special test-specific analyzer
+        // references.  So we can't assert we *only* have that type.
+        //
+        // We can *firmly* state though that we should never get an AnalyzerFileReference or IsolatedAnalyzerReference
+        // as that would mean we were not properly getting the real analyzer references produced by the serialized
+        // system.
+        Contract.ThrowIfTrue(serializedReferences.Any(r => r is AnalyzerFileReference));
```

Comment:
>AnalyzerFileReference [](http://example.com/codeflow?start=64&length=21)

maybe just one ThrowIfTrue: 

r => r is AnalyzerFileReference or IsolatedAnalyzerReference

Reply:
i liked having two checks so that if one throws we know which one it is.